### PR TITLE
further ameliorate the volume attach/detach polling issues

### PR DESCRIPTION
**Pull request by boris-dimitrov**
reduce probability of collisions on devnode

slow down the wait-for-volume-to-attach loop with exponential delays
and pid-based jitter so it doesn't run up against api rate limits

### CODE REVIEW

**Comment by boris-dimitrov** [comment]
@kislyuk thanks for your attention to this... the rate limit is now affecting other biohub users in addition to idseq

**Comment by boris-dimitrov** [comment]
cc @yunfangjuan @jameshowardwang appreciate your eyes on this, i did limited testing on the bash prompt

File: aegea/batch.py [hunk 1]
```diff
@@ -25,16 +25,42 @@
...
+
+# let each process start trying from a different /dev/xvd{letter}
+let pid=$$
+alphabet=$(echo {a..z})
+device_letters="${alphabet:pid%27*2} ${alphabet:0:pid%27*2}"
+
+# when N processes compete, for every success there can be N-1 failures; so the appropriate number of retries is O(N^2)
+# let us size this for 10 competitors
+let delay=2
+for try in {1..100}; do if [[ $try == 9 ]]; then echo "Unable to mount $vid on $devnode"; exit 1; fi; if [[ $try -gt 1 ]]; then sleep $delay; let delay=$delay*2; fi; for devlet in $device_letters; do devnode=/dev/xvd${devlet}; [[ -e $devnode ]] || break; done; aws ec2 attach-volume --instance-id $(echo "$iid" | jq -r .instanceId) --volume-id $vid --device $devnode || continue; break; done
```
Comment (boris-dimitrov) [inline aegea/batch.py]: yikes caught a small bug here

File: aegea/batch.py [hunk 2]
```diff
@@ -25,16 +25,42 @@
...
+
+# let each process start trying from a different /dev/xvd{letter}
+let pid=$$
+alphabet=$(echo {a..z})
+device_letters="${alphabet:pid%27*2} ${alphabet:0:pid%27*2}"
+
+# when N processes compete, for every success there can be N-1 failures; so the appropriate number of retries is O(N^2)
+# let us size this for 10 competitors
+let delay=2
+for try in {1..100}; do if [[ $try == 100 ]]; then echo "Unable to mount $vid on $devnode"; exit 1; fi; if [[ $try -gt 1 ]]; then sleep $delay; let delay=$delay*2; fi; for devlet in $device_letters; do devnode=/dev/xvd${devlet}; [[ -e $devnode ]] || break; done; aws ec2 attach-volume --instance-id $(echo "$iid" | jq -r .instanceId) --volume-id $vid --device $devnode || continue; break; done
```
Comment (boris-dimitrov) [inline aegea/batch.py]: I am also considering this version here:
```
# let each process start trying from a different /dev/xvd{letter}
let pid=$$

# when N processes compete, for every success there can be N-1 failures; so the appropriate number of retries is O(N^2)
# let us size this for 10 competitors
let delay=2+$pid%5
for try in {1..100}; do if [[ $try == 100 ]]; then echo "Unable to mount $vid on $devnode"; exit 1; fi; if [[ $try -gt 1 ]]; then sleep $delay; fi; devices=$(echo /dev/xvd* /dev/xvd{a..z} /dev/xvd{a..z} | sed 's= =\n=g' | sort | uniq -c | sort -n | grep ' 2 ' | awk '{print $2}'); let devcnt=${#devices}/10+1; let ind=$pid%devcnt; devnode=${devices:10*$ind:9};  aws ec2 attach-volume --instance-id $(echo "$iid" | jq -r .instanceId) --volume-id $vid --device $devnode || continue; break; done
```

**Comment by kislyuk** [comment]
Wow, this is impressively complex. I'm sorry the process of deconflicting these EBS volume mounts is so cumbersome. Unfortunately I'm not aware of any great alternatives on AWS Batch right now.

**Comment by kislyuk** [comment]
Released in v2.0.0 (this is not really a major release; I'm bumping the major version because my release management script has a bug when bumping multi-digit version components). I did *not* test this volume management shellcode myself. I'm relying on you guys to tell me if it works :)

**Comment by boris-dimitrov** [comment]
Thanks Andrey.   We are about to put this in production.   Will submit any bug fixes to you promptly as we bugs are discovered.

**Pull request merged**